plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/glue/TrinoGlueCatalog.java

```diff
@@ -351,6 +365,107 @@ public List<SchemaTableName> listTables(ConnectorSession session, Optional<Strin
+                    if (isTrinoMaterializedView(tableType, tableParameters)) {
+                        Optional<String> comment = decodeMaterializedViewData(table.getViewOriginalText()).getComment();
+                        unfilteredResult.add(RelationCommentMetadata.forTable(name, comment));
+                    }
+                    else if (isPrestoView(tableParameters)) {
+                        Optional<String> comment = ViewReaderUtil.PrestoViewReader.decodeViewData(table.getViewOriginalText()).getComment();
+                        unfilteredResult.add(RelationCommentMetadata.forTable(name, comment));
+                    }
+                    else if (isRedirected.test(name)) {
+                        unfilteredResult.add(RelationCommentMetadata.forRedirectedTable(name));
+                    }
+                    else if (!isIcebergTable(tableParameters)) {
+                        // This can be e.g. Hive, Delta table, a Hive view, etc. Would be returned by listTables, so do not skip it
+                        unfilteredResult.add(RelationCommentMetadata.forTable(name, Optional.empty()));
+                    }
+                    else {
+                        String metadataLocation = tableParameters.get(METADATA_LOCATION_PROP);
+                        String metadataValidForMetadata = tableParameters.get(TRINO_TABLE_METADATA_INFO_VALID_FOR);
+                        if (metadataValidForMetadata != null && metadataValidForMetadata.equals(metadataLocation)) {
+                            Optional<String> comment = Optional.ofNullable(tableParameters.get(TABLE_COMMENT));
+                            unfilteredResult.add(RelationCommentMetadata.forTable(name, comment));
+                        }
+                        else {
+                            unprocessed.put(name, table);
+                            if (unprocessed.size() >= PER_QUERY_CACHE_SIZE) {
+                                extractTableComments(session, unprocessed, relationFilter, filteredResult::add);
+                                unprocessed.clear();
+                            }
+                        }
+                    }
+                });
+
+        if (!unprocessed.isEmpty()) {
+            extractTableComments(session, unprocessed, relationFilter, filteredResult::add);
+        }
+
+        List<RelationCommentMetadata> unfilteredResultList = unfilteredResult.build();
+        Set<SchemaTableName> availableNames = relationFilter.apply(unfilteredResultList.stream()
+                .map(RelationCommentMetadata::name)
+                .collect(toImmutableSet()));
+
+        return Optional.of(Stream.concat(
+                        unfilteredResultList.stream()
+                                .filter(commentMetadata -> availableNames.contains(commentMetadata.name())),
+                        filteredResult.build().stream())
+                .iterator());
```

Comment:
We're doing a lot of eager work here. It doesn't matter yet because all the streams/iterators are collected to Lists farther up but if we switch the engine to do more of that lazily we'll want to come back to this.

Reply:
agreed. however, i didn't see a clean way to to this somewhat non-trivial logic, where we want to filter tables _after_ capturing comments for cases where we have comment information at hand, and we want to filter tables _before_ going to file system. Doing it more lazy would mean we consult access control more times, which is going to be more expensive.
(previous table_comments implementation would gather even more state in-memory)